os/mm/mm_heap/mm_memalign.c

```diff
@@ -108,137 +130,146 @@ FAR void *mm_memalign(FAR struct mm_heap_s *heap, size_t alignment, size_t size)
+	newsize = MM_ALIGN_UP(size + SIZEOF_MM_ALLOCNODE);
 
-	/* Get the node associated with the allocation and the next node after
-	 * the allocation.
-	 */
+	/* We need to hold the MM semaphore while we muck with the nodelist. */
 
-	node = (FAR struct mm_allocnode_s *)(rawchunk - SIZEOF_MM_ALLOCNODE);
+	mm_takesemaphore(heap);
 
-#ifdef CONFIG_DEBUG_MM_HEAPINFO
-		heapinfo_subtract_size(heap, node->pid, node->size);
-		heapinfo_update_total_size(heap, ((-1) * (node->size)), node->pid);
-#endif
-	/* Find the aligned subregion */
+	/* Get the location in the node list to start the search
+	 * by converting the request size into a nodelist index.
+	 */
 
-	alignedchunk = (rawchunk + mask) & ~mask;
+	ndx = mm_size2ndx(newsize);
 
-	/* Check if there is free space at the beginning of the aligned chunk */
+	/* Search for a large enough chunk in the list of nodes.
+	 * mm_nodelist is an array of lists. The array is arranged in ascending order of size.
+	 * Each list is ordered by size in a descending order.
+	 * If this list does not have free nodes whose size is large enough
+	 * to accomodate the requested size, it will fail due to no more space.
+	 */
+	for (; ndx < MM_NNODES; ndx++) {
+		node = heap->mm_nodelist[ndx].flink;
+		if (!(node && node->size >= newsize)) {
+			/* If the list at this index is empty or if the size of first node
+			 * in the list is less than the required size, then go to next index.
+			 */
+			continue;
+		}
 
-	if (alignedchunk != rawchunk) {
-		FAR struct mm_allocnode_s *newnode;
-		FAR struct mm_allocnode_s *next;
-		size_t precedingsize;
+		FAR struct mm_freenode_s *prev = &heap->mm_nodelist[ndx];
 
-		/* Get the node the next node after the allocation. */
+		/* Traverse the list until the end or until the node size is less than required size */
+		for ( ; node && node->size >= newsize; prev = node, node = node->flink);
 
-		next = (FAR struct mm_allocnode_s *)((char *)node + node->size);
+		/* If you hit the end of list, then set the node to prev */
+		if (!node) {
+			node = prev;
+		}
 
-		/* Make sure that there is space to convert the preceding mm_allocnode_s
-		 * into an mm_freenode_s.  I think that this should always be true
-		 */
+		/* Now, traverse the list in reverse direction, towards bigger size nodes */
+		for ( ; node; node = node->blink) {
+			alignchunk = (FAR struct mm_allocnode_s *)(((size_t)node + SIZEOF_MM_ALLOCNODE + mask) & ~mask);
+			size_t alignsize = (size_t)alignchunk - (size_t)node + size;
+			size_t remainsize = (size_t)alignchunk - SIZEOF_MM_ALLOCNODE - (size_t)node;
+
+			/* We found a suitable node if node size is more than required size after alignment and
+			 * if the remaining bytes before the alignment point is either zero or bigger than freenode.
+			 */
+			if (node->size >= alignsize && (remainsize == 0 || remainsize >= SIZEOF_MM_FREENODE)) {
+				found_align = true;
+				break;
+			}
+		}
 
-		DEBUGASSERT(alignedchunk >= rawchunk + 8);
+		if (found_align) {
```

Comment:
If we cannot find the align one, `found_align` remains false. In this case, we will return NULL for failure.
So I think, if we modify this line with false checking, we don't need to check `found_align` in line 201.
```
if (found_align == false) {
    return NULL;
}
```